zarr/storage.py

```diff
@@ -1144,9 +1147,28 @@ def dir_path(self, path=None):
     def listdir(self, path=None):
         dir_path = self.dir_path(path)
         try:
-            out = sorted(p.rstrip('/').rsplit('/', 1)[-1]
-                         for p in self.fs.ls(dir_path, detail=False))
-            return out
+            children = sorted(p.rstrip('/').rsplit('/', 1)[-1]
+                              for p in self.fs.ls(dir_path, detail=False))
+            if self.key_separator != "/":
+                return children
+            else:
+                if array_meta_key in children:
+                    # special handling of directories containing an array to map nested chunk
```

Comment:
@martindurant: can you comment if this matches your expectations?

Reply:
@martindurant : any thoughts?

Reply:
OK, now I have reminded myself  what it's doing, and I suppose this is reasonable. As to it's accuracy in all cases, I can only guess it looks right.

It makes you wonder how useful `listdir` actually if for the case of `fsspec`. I suppose it remains true that a user might want to start exploring a dataset at some high-level group and descend to a specific array, yet make sure they never list the entire set of files (which could be expensive).

Reply:
Thanks, @martindurant. I'm going to take that as a :thumbsup: and get this into a 2.8.1 release. It seems that this may be something that needs a re-evaluation down the line.